cluster-api/apis/cluster/types.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster contains types to represent Kubernetes cluster configuration.
+package cluster // import "k8s.io/kube-deploy/cluster-api/apis/cluster"
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// Cluster is an API object representing a cluster's control-plane
+// configuration and status.
+type Cluster struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	Spec   ClusterSpec
+	Status ClusterStatus
+}
+
+type ClusterSpec struct {
```

Comment:
@pipejakob @krousey 

One thing to consider is how we want to model deletion with this API. For example, what does the customer to do the API represent deletion. How do we want to represent things that are in the process of deletion/cleanup but are not done yet?

Reply:
When deleted, and API object doesn't go away immediately. Deletion can be deferred until all related resources are cleaned up.

Reply:
Hmmmm... I suppose since a cluster deletion can be done directly (as opposed to via a reconciler), it is not as big of a deal for cluster deletion to have a way to represent "try-to-delete-this-cluster".

However, I do find the machine API confusing regarding deletion. It has a deletion message so it seems that the API object is supposed to stick around during the deletion. However, I do not understand how the user updates that API object in a way that tell the machine reconciler to try deleting the machine. @pipejakob 

If we are ever going with a method to represent "try-to-delete-this-cluster" for the cluster, it would be good to match that with the machine API.

Reply:
From the Machines API:

    // It is recommended, but not required, that provider-specific controllers add
    // finalizers to Machine objects so that they can be triggered on deletion to
    // release the necessary external resources, reporting any errors encountered.

The idea is that a client that wants to delete a Machine will actually just use the REST API to perform DELETE on the specific Machine resource they want deleted. If no one has registered a finalizer for the object, then it will disappear immediately, but the cloud controller responsible for releasing the underlying resources should be watching all Node and Machine objects. If it receives an event for the Machine deletion, it can act immediately on it. If the controller was disconnected from the apiserver and missed that event, it can still be querying for Node objects who have had their OwnerRef specified and gone missing (this is meant to point to the Machine responsible for the Node), to basically constantly be looking for orphans that should be taken out of service.

If the controller wants to have even more control, as is suggested as a good pattern, then it can register a finalizer for every Machine object that it's responsible for so that when one is deleted, the object doesn't actually disappear until the controller finalizer code has run and unregistered itself. In this case, the client still calls DELETE on the resource, but there will be a delay to wait for any finalizers to run (so the controller can delete the VM) and unregister its finalizer when done, so that the apiserver can finally delete the object afterward. If the controller runs into any issues, like the cloud API it's trying to call returns an error, or it suddenly has authentication issues, then the Machine object still exists, and the controller can add error information to its MachineStatus.

If a human or any tooling tries to delete a Machine, and it cares about the success of that operation, it should continue to watch the object for any errors that appear in its status, or consider the operation successful when the delete finally succeeds.

More about finalizers: https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#finalizers

Reply:
This is what I was referring to. We can accomplish complete cluster deletion of resources tracked within the API through cascading deletion and owner references. For resources outside of the API, we can utilize finalizers to inform the API server when the actual API object can be erased from the store.